### PR TITLE
feat: Stream Feature View FCOS

### PR DESCRIPTION
**Pull request by kevjumba**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

### CODE REVIEW

File: protos/feast/core/StreamFeatureView.proto [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import "feast/core/Feature.proto";
+import "feast/core/DataSource.proto";
+
+message StreamFeatureView {
+    // User-specified specifications of this feature view.
+    StreamFeatureViewSpec spec = 1;
+    StreamFeatureViewMeta meta = 2;
+}
+
+// Next available id: 10
```
Comment (felixwang9817) [inline protos/feast/core/StreamFeatureView.proto:40]: nit: looks like it should be 17

File: protos/feast/core/StreamFeatureView.proto [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    // Batch/Offline DataSource where this view can retrieve offline feature data.
+    DataSource batch_source = 10;
+    // Streaming DataSource from where this view can consume "online" feature data.
+    DataSource stream_source = 11;
+
+    // Whether these features should be served online or not
+    bool online = 12;
+
+    UserDefinedFunction user_defined_function = 13;
```
Comment (felixwang9817) [inline protos/feast/core/StreamFeatureView.proto:80]: nit: add comment

File: protos/feast/core/StreamFeatureView.proto [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+message StreamFeatureViewMeta {
+    // Time where this Feature View is created
+    google.protobuf.Timestamp created_timestamp = 1;
+
+    // Time where this Feature View is last updated
+    google.protobuf.Timestamp last_updated_timestamp = 2;
+}
+
+message Aggregation {
```
Comment (felixwang9817) [inline protos/feast/core/StreamFeatureView.proto:100]: probably should be in its own file, similar to `Field.proto`

File: sdk/python/feast/stream_feature_view.py [hunk 4]
```diff
@@ -1,32 +1,112 @@
...
+from feast.protos.feast.core.StreamFeatureView_pb2 import (
+    StreamFeatureViewSpec as StreamFeatureViewSpecProto,
+)
+
+warnings.simplefilter("once", RuntimeWarning)
+
+SUPPORTED_STREAM_SOURCES = {"KafkaSource", "PushSource"}
+
+
+class Aggregation(abc.ABC):
```
Comment (felixwang9817) [inline sdk/python/feast/stream_feature_view.py:37]: nit: this should probably be in its own file, similar to `field.py`

File: sdk/python/feast/stream_feature_view.py [hunk 5]
```diff
@@ -1,32 +1,112 @@
...
+    """
+    NOTE: Feast-handled aggregations are not yet supported. This class provides a way to register user-defined aggregations.
+    """
+
+    column: str  # Column name of the feature we are aggregating.
```
Comment (felixwang9817) [inline sdk/python/feast/stream_feature_view.py:42]: nit: would prefer these comments be in an Attributes section, see e.g. `field.py`

File: sdk/python/feast/stream_feature_view.py [hunk 6]
```diff
@@ -1,32 +1,112 @@
...
+    function: str  # Provided built in aggregations sum, max, min, count mean
+    time_windows: List[str]  # The time window. Example ["1h", "24h"]
```
Comment (felixwang9817) [inline sdk/python/feast/stream_feature_view.py:44]: I think we should be using `timedelta` here? (see spec)

File: sdk/python/feast/stream_feature_view.py [hunk 7]
```diff
@@ -1,32 +1,112 @@
...
+        return aggregation
+
+    def __eq__(self, other):
+        if not isinstance(other, Aggregation):
+            raise TypeError("Comparisons should only involve Aggregations.")
+
+        if (
+            self.column != other.column
+            or self.function != other.function
+            or self.time_windows != other.time_windows
```
Comment (felixwang9817) [inline sdk/python/feast/stream_feature_view.py:72]: might need to sort this list? depends on our exact assumptions, just wanted to call it out

File: sdk/python/tests/unit/test_feature_views.py [hunk 8]
```diff
@@ -78,4 +84,43 @@ def test_create_stream_feature_view():
...
+        timestamp_field="",
+        bootstrap_servers="",
+        message_format=AvroFormat(""),
+        topic="topic",
+        batch_source=FileSource(path="some path"),
+    )
+
+    sfv = StreamFeatureView(
+        name="test kafka stream feature view",
+        entities=["driver"],
```
Comment (felixwang9817) [inline sdk/python/tests/unit/test_feature_views.py:107]: nit: use actual entity instead of string

File: sdk/python/feast/feature_store.py [hunk 9]
```diff
@@ -1152,8 +1195,8 @@ def materialize(
 
         Examples:
             Materialize all features into the online store over the interval
-            from 3 hours ago to 10 minutes ago.
 
+            from 3 hours ago to 10 minutes ago.
```
Comment (felixwang9817) [inline sdk/python/feast/feature_store.py:1199]: nit: not sure why newline got added

File: sdk/python/feast/stream_feature_view.py [hunk 10]
```diff
@@ -35,17 +115,205 @@ def __init__(
...
+
+        if not super().__eq__(other):
+            return False
+
+        if (
+            self.mode != other.mode
+            or self.timestamp_field != other.timestamp_field
+            or self.udf.__code__.co_code != other.udf.__code__.co_code
+            or self.aggregations != other.aggregations
+            or self.timestamp_field != other.timestamp_field
```
Comment (felixwang9817) [inline sdk/python/feast/stream_feature_view.py:153]: nit: repeated

File: sdk/python/feast/stream_feature_view.py [hunk 11]
```diff
@@ -35,17 +115,205 @@ def __init__(
...
+        if self.batch_source:
+            batch_source_proto = self.batch_source.to_proto()
+            batch_source_proto.data_source_class_type = f"{self.batch_source.__class__.__module__}.{self.batch_source.__class__.__name__}"
+
+        stream_source_proto = None
+        if self.stream_source:
+            stream_source_proto = self.stream_source.to_proto()
+            stream_source_proto.data_source_class_type = f"{self.stream_source.__class__.__module__}.{self.stream_source.__class__.__name__}"
+
+        aggregation_proto_lst = []
```
Comment (felixwang9817) [inline sdk/python/feast/stream_feature_view.py:183]: nit: list?

File: sdk/python/tests/integration/registration/test_registry.py [hunk 12]
```diff
@@ -299,6 +301,62 @@ def location_features_from_push(inputs: pd.DataFrame) -> pd.DataFrame:
...
+        udf=simple_udf,
+        tags={},
+    )
+
+    project = "project"
+
+    # Register Feature View
+    test_registry.apply_feature_view(sfv, project)
+
+    feature_views = test_registry.list_stream_feature_views(project)
```
Comment (felixwang9817) [inline sdk/python/tests/integration/registration/test_registry.py:344]: nit: maybe name this stream_feature_views?

**Review by felixwang9817 — COMMENTED** [review]
some initial comments, looks good overall!

File: protos/feast/core/StreamFeatureView.proto (hunk 1 above)
Comment (kevjumba) [reply protos/feast/core/StreamFeatureView.proto:40]: fixed

File: protos/feast/core/StreamFeatureView.proto (hunk 2 above)
Comment (kevjumba) [reply protos/feast/core/StreamFeatureView.proto:80]: fixed

File: sdk/python/feast/stream_feature_view.py (hunk 6 above)
Comment (kevjumba) [reply sdk/python/feast/stream_feature_view.py:44]: done

File: sdk/python/feast/stream_feature_view.py (hunk 7 above)
Comment (kevjumba) [reply sdk/python/feast/stream_feature_view.py:72]: good catch!

File: sdk/python/feast/stream_feature_view.py (hunk 11 above)
Comment (kevjumba) [reply sdk/python/feast/stream_feature_view.py:183]: fixed

File: sdk/python/feast/stream_feature_view.py (hunk 4 above)
Comment (kevjumba) [reply sdk/python/feast/stream_feature_view.py:37]: fixed

File: sdk/python/tests/unit/test_feature_views.py [hunk 13]
```diff
@@ -2,12 +2,14 @@
 
 import pytest
 
-from feast import PushSource
+from feast import Entity, Field, PushSource
```
Comment (felixwang9817) [inline sdk/python/tests/unit/test_feature_views.py:5]: nit: we want to do `from feast.entity import Entity` etc. for all internal files

File: sdk/python/feast/feature_store.py [hunk 14]
```diff
@@ -488,16 +510,25 @@ def _make_inferences(
...
+        update_data_sources_with_inferred_event_timestamp_col(
+            [view.batch_source for view in sfvs_to_update], self.config
+        )
+
         # New feature views may reference previously applied entities.
         entities = self._list_entities()
         update_feature_views_with_inferred_features_and_entities(
             views_to_update, entities + entities_to_update, self.config
         )
+        update_feature_views_with_inferred_features_and_entities(
```
Comment (felixwang9817) [inline sdk/python/feast/feature_store.py:522]: this inference will break for SFVs as it relies on the underlying batch source; instead we should be doing something similar to ODFV schema inference

I think we can do that in a future PR; for now maybe don't run this inference here, and just add a warning or something in SFV if a schema isn't specified?

File: sdk/python/feast/feature_store.py [hunk 15]
```diff
@@ -549,13 +581,15 @@ def _plan(
             desired_repo_contents.feature_views,
             desired_repo_contents.on_demand_feature_views,
             desired_repo_contents.request_feature_views,
+            desired_repo_contents.stream_feature_views,
```
Comment (felixwang9817) [inline sdk/python/feast/feature_store.py:584]: given that you're changing the `feast apply` flow I think we should also add a test that SFVs can be applied properly - maybe something similar to `test_feature_service_apply.py`?

File: sdk/python/feast/stream_feature_view.py [hunk 16]
```diff
@@ -35,17 +70,195 @@ def __init__(
...
+    mode: Optional[str] = "spark",  # Mode of ingestion/transformation
+    timestamp_field: Optional[str] = "",  # Timestamp for aggregation
+):
+    """
+    Creates an StreamFeatureView object with the given user function as udf.
+    """
+
+    def mainify(obj):
+        # Needed to allow dill to properly serialize the udf. Otherwise, clients will need to have a file with the same
+        # name as the original file defining the ODFV.
```
Comment (felixwang9817) [inline sdk/python/feast/stream_feature_view.py:240]: nit: SFV instead of ODFV

File: sdk/python/feast/feature_store.py (hunk 15 above)
Comment (kevjumba) [reply sdk/python/feast/feature_store.py:584]: Added test.

File: sdk/python/feast/feature_store.py (hunk 14 above)
Comment (kevjumba) [reply sdk/python/feast/feature_store.py:522]: Added as a future pr, also added warning

File: sdk/python/feast/aggregation.py [hunk 17]
```diff
@@ -0,0 +1,76 @@
+import abc
+from datetime import timedelta
+from typing import List, Union
+
+from google.protobuf.duration_pb2 import Duration
+
+from feast.protos.feast.core.Aggregation_pb2 import Aggregation as AggregationProto
+
+
+class Aggregation(abc.ABC):
```
[the remark this answers is not in the document]
Comment (felixwang9817) [reply sdk/python/feast/aggregation.py:10]: hm don't think was removed @kevjumba

**Review by felixwang9817 — APPROVED** [review]
/lgtm

**Pull request merged**